File: packages/seismograph/ios/SeismographModule.swift
import ExpoModulesCore
import CoreMotion

public class SeismographModule: Module {
  private let motionManager = CMMotionManager()

  public func definition() -> ModuleDefinition {
    Name("SeismographModule")

    Events("onReading")

    Function("startMonitoring") { (intervalMs: Double) in
      guard self.motionManager.isAccelerometerAvailable else { return }
      self.motionManager.accelerometerUpdateInterval = intervalMs / 1000.0
      self.motionManager.startAccelerometerUpdates(to: .main) { [weak self] data, error in
        guard let data = data, error == nil else { return }
        self?.sendEvent("onReading", [
          "x": data.acceleration.x * 9.81,
          "y": data.acceleration.y * 9.81,
          "z": data.acceleration.z * 9.81,
          "timestamp": Date().timeIntervalSince1970 * 1000,
        ])
      }
    }

    Function("stopMonitoring") {
      self.motionManager.stopAccelerometerUpdates()
    }
  }
}

File: packages/seismograph/android/src/main/java/com/quakelink/seismograph/SeismographModule.kt
package com.quakelink.seismograph

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class SeismographModule : Module(), SensorEventListener {
  private var sensorManager: SensorManager? = null
  private var accelerometer: Sensor? = null
  private var intervalUs: Int = 50_000 // 50ms → 20 Hz

  override fun definition() = ModuleDefinition {
    Name("SeismographModule")

    Events("onReading")

    Function("startMonitoring") { intervalMs: Double ->
      intervalUs = (intervalMs * 1000).toInt()
      val ctx = appContext.reactContext ?: return@Function
      sensorManager = ctx.getSystemService(Context.SENSOR_SERVICE) as SensorManager
      accelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
      sensorManager?.registerListener(this@SeismographModule, accelerometer, intervalUs)
    }

    Function("stopMonitoring") {
      sensorManager?.unregisterListener(this@SeismographModule)
    }
  }

  override fun onSensorChanged(event: SensorEvent) {
    if (event.sensor.type != Sensor.TYPE_ACCELEROMETER) return
    sendEvent(
      "onReading",
      mapOf(
        "x" to event.values[0].toDouble(),
        "y" to event.values[1].toDouble(),
        "z" to event.values[2].toDouble(),
        "timestamp" to System.currentTimeMillis().toDouble(),
      )
    )
  }

  override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
}
